app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt

```diff
@@ -85,68 +87,180 @@ class WebViewPresenterImpl @Inject constructor(
-                oldUrl?.host != url?.host ||
-                oldUrl?.port != url?.port
-            ) {
-                view.loadUrl(
-                    url = url.toString().toUri()
-                        .buildUpon()
-                        .appendQueryParameter("external_auth", "1")
-                        .build()
-                        .toString(),
-                    keepHistory = oldUrlForServer == urlForServer,
-                    openInApp = url?.baseIsEqual(baseUrl) ?: false,
-                )
-            }
+    private suspend fun handleUrlState(
+        urlState: UrlState,
+        path: String?,
+        shouldConsumePath: Boolean,
+        isNewServer: Boolean,
+    ) {
+        when (urlState) {
+            is UrlState.HasUrl -> handleHasUrlState(
+                urlState = urlState,
+                path = path,
+                shouldConsumePath = shouldConsumePath,
+                isNewServer = isNewServer,
+            )
+            UrlState.InsecureState -> handleInsecureState()
         }
     }
 
+    private suspend fun handleHasUrlState(
+        urlState: UrlState.HasUrl,
+        path: String?,
+        shouldConsumePath: Boolean,
+        isNewServer: Boolean,
+    ) {
+        if (shouldConsumePath) {
+            loadInitialUrl(
+                baseUrl = urlState.url,
+                path = path,
+                isNewServer = isNewServer,
+            )
+        } else {
+            // TODO not sure if we need to handle the path here, since if we end up into this
+            //  we would have already consumed the path and we are changing to another base URL
+            //  like going from internal to external.
+            loadServerUrl(
+                serverUrl = urlState.url,
+                isNewServer = isNewServer,
+            )
+        }
+    }
+
+    /**
+     * Loads the initial URL, optionally applying a path override.
+     *
+     * @param baseUrl the base server URL
+     * @param path optional path to append (ignored if starts with "entityId:")
+     * @param isNewServer whether this is a new server (affects history behavior)
+     */
+    private suspend fun loadInitialUrl(baseUrl: URL?, path: String?, isNewServer: Boolean) {
+        val urlToLoad = if (path != null && !path.startsWith("entityId:")) {
+            UrlUtil.handle(baseUrl, path)
+        } else {
+            baseUrl
+        }
+
+        urlToLoad?.let {
+            loadUrlInWebView(
+                url = it,
+                keepHistory = !isNewServer,
+                openInApp = it.baseIsEqual(baseUrl),
+            )
+        } ?: Timber.w("Url is null")
+    }
+
+    /**
+     * Loads a server URL directly without path modification.
+     *
+     * @param serverUrl the URL to load
+     * @param isNewServer whether this is a new server (affects history behavior)
+     */
+    private suspend fun loadServerUrl(serverUrl: URL?, isNewServer: Boolean) {
```

Comment:
Why is this a different function and not simply `loadInitialUrl` with a null path?